### PR TITLE
Update interface Function present as dependency in pass `array_op`

### LINKED ISSUE #1194: Bug in pass `subroutine_from_function`
While trying to compile `minpack/examples/example_primes.f90`, I came across
```fortran
subroutine find_fit(expr)
interface
    function expr() result(y)
    implicit none
    real :: y(1)
    end function
end interface
end subroutine
    
program main
    implicit none
    call find_fit(expression)
    contains
    function expression() result(y)
    real :: y(1)
    end function
end program
```
>LFortran

```console
$ lfortran a.f90
ASR verify pass error: ASR verify: Var::m_v `expression` cannot point outside of its symbol table
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/home/pranavchiku/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/pranavchiku/lfortran/src/bin/lfortran.cpp", line 1934, in ??
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/home/pranavchiku/lfortran/src/bin/lfortran.cpp", line 862, in ??
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/home/pranavchiku/lfortran/src/lfortran/fortran_evaluator.cpp", line 338, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/pranavchiku/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 6753, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass_manager.apply_passes(al, &asr, pass_options, diagnostics);
  File "/home/pranavchiku/lfortran/src/libasr/pass/pass_manager.h", line 215, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    _apply_passes(al, asr, _passes, pass_options, diagnostics);
LCompilersException: Verify failed
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1194.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
This logic resolves the bug for the given issue but then generates the following log, 
```console
$ lfortran a.f90
; ModuleID = 'LFortran'
source_filename = "LFortran"

%array = type { float*, i32, %dimension_descriptor*, i1, i32 }
%dimension_descriptor = type { i32, i32, i32 }

define void @find_fit(float ()* %expr) {
.entry:
  br label %return

return:                                           ; preds = %.entry
  ret void
}

declare float @expr()

define float @expression() {
.entry:
  %y = alloca %array, align 8
  %0 = getelementptr %array, %array* %y, i32 0, i32 1
  store i32 0, i32* %0, align 4
  %1 = getelementptr %array, %array* %y, i32 0, i32 2
  %2 = alloca i32, align 4
  store i32 1, i32* %2, align 4
  %3 = load i32, i32* %2, align 4
  %4 = alloca %dimension_descriptor, i32 %3, align 8
  %5 = getelementptr %array, %array* %y, i32 0, i32 4
  store i32 1, i32* %5, align 4
  store %dimension_descriptor* %4, %dimension_descriptor** %1, align 8
  %6 = load %dimension_descriptor*, %dimension_descriptor** %1, align 8
  %7 = getelementptr inbounds %dimension_descriptor, %dimension_descriptor* %6, i32 0
  %8 = getelementptr %dimension_descriptor, %dimension_descriptor* %7, i32 0, i32 0
  %9 = getelementptr %dimension_descriptor, %dimension_descriptor* %7, i32 0, i32 1
  %10 = getelementptr %dimension_descriptor, %dimension_descriptor* %7, i32 0, i32 2
  store i32 1, i32* %8, align 4
  store i32 1, i32* %9, align 4
  store i32 1, i32* %10, align 4
  %11 = alloca i32, align 4
  store i32 1, i32* %11, align 4
  %12 = getelementptr %array, %array* %y, i32 0, i32 0
  %13 = load i32, i32* %11, align 4
  %14 = alloca float, i32 %13, align 4
  store float* %14, float** %12, align 8
  br label %return

return:                                           ; preds = %.entry
  %15 = load %array, %array* %y, align 8
  ret %array %15
}

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_set_argv(i32 %0, i8** %1)
  call void @find_fit(float ()* @expression)
  ret i32 0
}

declare void @_lpython_set_argv(i32, i8**)
code generation error: asr_to_llvm: module failed verification. Error:
Function return type does not match operand type of return inst!
  ret %array %15
 float
```

**Comment by Pranavchiku** [comment]
I guess a better way will be to convert the function `expr` also to a subroutine if Function `expression` is present as a dependency.

**Comment by certik** [comment]
Thanks for trying to figure it out! If you don't know how to continue, we can meet and work on this together. I think we are very close, this is one of the last bugs.

**Comment by Pranavchiku** [comment]
Yeah, perfect!

File: tests/array12.f90 [hunk 1]
```diff
@@ -0,0 +1,17 @@
+subroutine find_fit(expr)
+    interface
+        function expr() result(y)
+        implicit none
+        real :: y(1)
+        end function
+    end interface
```
Comment (certik) [inline tests/array12.f90:7]: In this PR, I think we have to convert this interface function into a subroutine, then things should start working.

**Review by certik — APPROVED** [review]
Otherwise this looks good. It seems it works now, I tested it locally.

**Comment by certik** [comment]
Thanks!

**Pull request merged**